tests/bluetooth/bsim_bt/bsim_test_l2cap_ecred/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,540 @@
+	.recv = chan_recv_cb,
+	.sent = chan_sent_cb,
+	.connected = chan_connected_cb,
+	.disconnected = chan_disconnected_cb,
+	.status = chan_status_cb,
+	.released = chan_released_cb,
+	.reconfigured = chan_reconfigured_cb,
+};
+
+static void connect_num_channels(uint8_t num_l2cap_channels)
+{
+	struct channel *chan = NULL;
+
+	#define L2CAP_ECRED_CHAN_MAX 5
+	struct bt_l2cap_chan *allocated_channels[L2CAP_ECRED_CHAN_MAX] = { NULL };
+	#undef L2CAP_ECRED_CHAN_MAX
+	int err = 0;
+	for (int i = 0; i < num_l2cap_channels; i++) {
+		chan = get_free_channel();
+		if (!chan) {
+			FAIL("failed, chan not free");
+			return;
+		}
+		chan->le.chan.ops = &l2cap_ops;
+		chan->le.rx.mtu = DATA_MTU;
+		chan->le.rx.mps = DATA_MPS;
+		allocated_channels[i] = &chan->le.chan;
+	}
+
+	err = bt_l2cap_ecred_chan_connect(default_conn, allocated_channels, servers[0].psm);
+	if (err) {
+		FAIL("can't connect ecred %d ", err);
+	}
+}
+
+static void disconnect_all_channels(void)
+{
+	int err = 0;
+
+	for (int i = 0; i < ARRAY_SIZE(channels); i++) {
+		if (channels[i].in_use) {
+			LOG_DBG("Disconnecting channel: %d)", channels[i].chan_id);
+			err = bt_l2cap_chan_disconnect(&channels[i].le.chan);
+			if (err) {
+				LOG_DBG("can't disconnnect channel (err: %d)", err);
+			}
+			channels[i].in_use = false;
+		}
+	}
+}
+
+static int accept(struct bt_conn *conn, struct bt_l2cap_chan **l2cap_chan)
+{
+	struct channel *chan;
+
+	chan = get_free_channel();
+	if (!chan) {
+		return -ENOMEM;
+	}
+
+	chan->le.chan.ops = &l2cap_ops;
+	chan->le.tx.mtu = DATA_MTU;
+	chan->le.rx.mtu = DATA_MTU;
+
+	*l2cap_chan = &chan->le.chan;
+
+	return 0;
+}
+
+static struct bt_l2cap_server *get_free_server(void)
+{
+	for (int i = 0U; i < SERVERS; i++) {
+		if (servers[i].psm) {
+			continue;
+		}
+
+		return &servers[i];
+	}
+
+	return NULL;
+}
+
+static void register_l2cap_server(void)
+{
+	struct bt_l2cap_server *server;
+
+	server = get_free_server();
+	if (!server) {
+		FAIL("Failed to get free server");
+		return;
+	}
+
+	server->accept = accept;
+	server->psm = 0;
+
+	if (bt_l2cap_server_register(server) < 0) {
+		FAIL("Failed to get free server");
+		return;
+	}
+
+	LOG_DBG("L2CAP server registered, PSM:0x%X", server->psm);
+}
+
+static void connected(struct bt_conn *conn, uint8_t conn_err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (conn_err) {
+		FAIL("Failed to connect to %s (%u)", addr, conn_err);
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+		return;
+	}
+
+	default_conn = bt_conn_ref(conn);
+	LOG_DBG("%s", addr);
+
+	SET_FLAG(is_connected);
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	LOG_DBG("%s (reason 0x%02x)", addr, reason);
+
+	if (default_conn != conn) {
+		FAIL("Conn mismatch disconnect %s %s)", default_conn, conn);
+		return;
+	}
+
+	bt_conn_unref(default_conn);
```

Comment:
If you want to use the same disconnected callback for central and peripheral, keep in mind that only the central should call `bt_conn_unref`.